C API extensions

samansmink opened the pull request:
This PR:
- Reworks the DuckDB C API header `duckdb.h` to use code-generation
- Adds a new C Extension API  based on a struct-of-functionpointers design similar to [sqlite](https://github.com/cloudmeter/sqlite/blob/master/sqlite3ext.h) 

## Why?
The C Extension API solves 3 major problems with extensions:

### Linking DuckDB into every extension
Currently every extension has a (part of) DuckDB linked into it statically. When an extension's init function is called, a pointer to the DuckDB instance is passed to the extension code. However, to interact with 
this pointer to the instance in any meaningful way, requires to be able to call into DuckDB code. There are two ways this can be achieved when extending the C++ API: Firstly, by relying on the RTLD_GLOBAL flag. This flag
can be set when dlopening binaries to ensure the dlopened extension can resolve its missing symbols against those in the parent process, i.e. DuckDB. Unfortunately we can't always control whether DuckDB itself is opened with RTLD global: this would require us to hack this into every client (because the default on MacOS is RTLD_LOCAL). The second way is what we currently do: statically link a copy of DuckDB into every extension. While this is super wasteful with regards to the binary size, it turns out to be quite effective in distributing extensions to many platforms.

Now using the extension C API implementation described here, this problem would be circumvented completely: the extension can call into duckdb using only function pointers: no linking required there! This makes the extensions really small! for example the extension in my demo branch compiles to a 17K binary.

### Rebuilding extensions for new DuckDB versions
C++ API Extension are now tied tightly to the DuckDB version: Because we make no guarantees about C++ API stability and also because of how we link DuckDB into every extension binary, all extensions need to be recompiled for new versions of DuckDB. With C api extensions, this is no longer the case. We can simply build the extension once, and as long as DuckDB supports that version of the C API, the extension binary will be loadable.

### Requiring C++ for writing your extensions
Cleanly writing extensions in C, Rust, Go will be possible building on top of the C API. By using the "passing a c struct" method from SQLite, we remove the need to link with duckdb during build time whatsoever, making things much simpler, allowing extension to be created purely using the extension header file.

## Implementation 
### Overview
The basic idea of this is that instead of relying on dynamic or static linking to call into DuckDB functions, the extension init function gets passed a struct containing all available functions. This would
look a little like:

```C++
typedef struct {
	// ...
	idx_t (*duckdb_data_chunk_get_size)(duckdb_data_chunk chunk);
	duckdb_vector (*duckdb_data_chunk_get_vector)(duckdb_data_chunk chunk, idx_t col_idx);
	// ...
} duckdb_ext_api_v1;
```

The code for a C Extension API based extension is very straightforward, due to some macros:

```C++
#include "duckdb_extension.h"

DUCKDB_EXTENSION_ENTRYPOINT(duckdb_connection connection, duckdb_extension_info info, duckdb_extension_access *access) {
    // Register extension stuff here
}
```
Checkout the capi_demo extension added in this PR for an example.

### Code generation
I've split up the C API across two files:
- `duckdb.h` is the original capi header file 
- `duckdb_extension.h` is the header that extensions need to include, it contains the struct-of-function-pointers and typedefs to the original extension names.

The code is generated from `json` files which are in `src/include/duckdb/main/capi/header_generation/`. There are 2 types of json files. **function definition files** define all possible functions that are in the C API. These can potentially be deprecated. **api definition files** contain the definition (and order) of the C API struct.

samansmink committed:
demo of a CAPI extension sqlite-style

samansmink committed:
add scripts for parsing to and from json for capi

samansmink committed:
fix various comment inconsistencies to improve parsing of comments

samansmink committed:
update scripts

samansmink committed:
codegen capi header

samansmink committed:
change paths to actual paths

samansmink committed:
fix syntax

samansmink committed:
add function jsons

samansmink committed:
revert earlier changes to capi header

samansmink committed:
apply various manual fixes to capi json files

samansmink committed:
extension capi sortof works, its a big mess though

samansmink committed:
cleanup extension capi

samansmink committed:
finished first version of capi extensions

samansmink commented:
@Mause @Mytherin @carlopi @Maxxen got a first version of c api, would love to get some feedback before spending more time here

rustyconover commented:
This looks fine from the perspective of someone experienced with the extension API. It won’t significantly affect API consumers from the Rust side.

I appreciate that this PR allows the extension API to evolve. I suggest adding the ability for DuckDB to support older extension API versions by passing multiple structures upon extension initialization. In the init function, can you extend it so DuckDB can send a list of supported extension API structures, and the extension can choose which version to use?

For example, `duckdb_function_set_error()` only works with table-returning functions, but it should be called `duckdb_table_function_set_error()`, given the existence of `duckdb_scalar_function_set_error()`. This could be addressed in a newer version of the API, where as existing extensions could just keep working like nothing ever changed.

Currently, we lack many features of the C++ extension API in the C API. Some examples are:

1. Interacting with duckdb_vector is difficult for extension writers. GenericExecutor and similar tools make this much simpler.
2. The C API cannot define new types and set aliases for them.
3. There is no way to perform predicate pushdown from the C API for table-returning functions.

Closing these gaps by making the extension API easier to extend would be beneficial.

Mytherin commented:
> Interacting with duckdb_vector is difficult for extension writers. GenericExecutor and similar tools make this much simpler.

The reason tools like `GenericExecutor` work well in C++ is due to templating/inlining - that unfortunately doesn't really work in C. As such you would incur function call overheads (that get optimized away in the C++ versions) - which would make such wrappers a lot slower than they would need to be.

I also think we should limit convenience wrappers in the C API and focus primarily on feature parity/performance. When the C API is sufficiently extended we can design wrappers in other languages that build on top of the C API (e.g. Rust/Go, maybe C++ itself) that build on top of this API and provide a better usability experience for writing extensions. Ideally few people would actually need to interact with the C API outside of the developers maintaining the wrappers at that point, making the convenience wrappers at that layer moot.

Maxxen commented:
I like the idea of maybe having multiple structs be returned as we could compartmentalize parts of the API. e.g. scalar and table functions could have their own API structs that can be evolved independently. Although I guess we would need a common "core" api for types and values anyway so maybe this just causes a bunch of unnecessary complexity.

samansmink commented:
@Maxxen 
> However, I wonder if it's not better to start from scratch with the extension API instead of having it be derived from the current C-API...? 

So for this demo I just added every function that is not deprecated to the v0 api struct. I think its a fair point that this would be and execellent point to reconsider if there are parts of the C CAPI we are not happy with and simply not include them in the struct. 

> table functions could have their own API structs that can be evolved independently.

That's an interesting idea, but it would indeed make things a bit more complex. Note also that even though we can then more easily change the apis, we can not make a lot of changes anyway, because we need to support both the before and after of every change we make, so we shouldn't be making lots of changes anyway. 

All in all I'm not super convinced that splitting the structs is worth the effort but let's discuss tomorrow!

@rustyconover
Thanks for the feedback! Passing an array of structs is an approach I had not considered yet, that does sound like an interesting option. Right now what i've done is use the binary footer that is added to every extension to store a semantic version number of the extension api. Before loading an extension, DuckDB will look up which version of the extension API the extension requires and pass it the struct it expects. I think that also solves what you want to achieve by passing an array of structs?

carlopi commented:
On Load, I would try hard to re-design the init function so that it takes parameters that are compatible between C++ and C API.
Given extension (up to this moment) are valid only for a given version, I think this is the right moment to do this, and rely on metadata to avoid that sane users will not mix and match.

I think it's better to have a single uniform init function that try to find a given function.
If we go with different functions, the main INIT function should still be declared and return failure in a way that is understood by the system (unsure how, since you can't throw, maybe adding a return parameter then)

Possibly for later, unsure what's the ideas here:
* unsure if CAPI extensions can be statically loaded at this moment, of what's the path there
* on Install, where are CAPI extension found ( either together with regular ones OR in a different subtree like capi/v0/$platform/$extension_name.duckdb_extension )
* on Install, where are CAPI extension saved locally (also here, either as a regular one OR as in a separate folder)

samansmink commented:
> On Load, I would try hard to re-design the init function so that it takes parameters that are compatible between C++ and C API.

Hmm technically c++ extensions can already use the C API init function I wrote here, they can just fetch the DatabaseInstance from the connection directly. Maybe we can indeed unify this right away. 


> I think it's better to have a single uniform init function that try to find a given function.
If we go with different functions, the main INIT function should still be declared and return failure in a way that is understood by the system (unsure how, since you can't throw, maybe adding a return parameter then)

i think i prefer letting duckdb handle the compatibility checking, but i agree that the hard dependency on having the metadata present is not ideal. I would say we have these options:
- add some macro magic to the `duckdb_extension.h` to verify the C api version in the `init` function extension side
- do some sort of handshake with a `_version` function that returns the C api extension version
- rely on the extension metadata as i've done in this PR

I went with #3, but i think either of these options could work


> unsure if CAPI extensions can be statically loaded at this moment, of what's the path there

we could and probably should have this. Adding it shouldn't be too hard though

> on Install, where are CAPI extension found .... on Install, where are CAPI extension saved locally

good point, i've left this to a future PR. it's implementation-wise completely separate from this PR, i think its just a matter of where to look for extensions for downloading and where to store/load them from

Mytherin reviewed (commented):
Thanks! Looks great - some comments from my side

samansmink committed:
small cleanup

samansmink committed:
just include the header, don't duplicate everything

samansmink committed:
first batch of review comments

samansmink committed:
refactor the extension c api to use duckdb_database instead of a connection

samansmink committed:
dynamic generation of c api struct

samansmink committed:
make single source of truth for CAPI_VERSION

samansmink committed:
Merge branch 'main' into c-api-extensions

samansmink committed:
push through first autogen changes

samansmink committed:
more merging of newly autogenerated code

samansmink committed:
more merging of newly autogenerated code

samansmink committed:
finished merge or autogen code from upstream c api changes

samansmink committed:
format

samansmink committed:
cleaner function comments with @

samansmink committed:
re-added accidentally borked functions from merge

samansmink committed:
fix copy error

samansmink committed:
minor clean up

samansmink committed:
more c api cleanup

samansmink committed:
fix duckdb_extension.h ordering

samansmink committed:
fix failing test due to error msg

samansmink committed:
cleanup

samansmink committed:
Merge branch 'main' into c-api-extensions

samansmink committed:
add capi_demo to ci

samansmink commented:
I think this is getting close. I think another review make sense here @Mytherin. I've updated the API definition files to include the functions that were added while this PR was open. and processed the review comments

Windows CI still has some issues which I fail to reproduce in my UTM VM for some reason. I'll try to fix that in the meantime.

Stuff changes since last review:
- All deprecated functions now have an `excluded_from_struct` param set to hide them from the v0 struct. This is something that we can only do now obiously and not when we start with the v1 versioned function pointer struct.
- the duckdb_extension.h header now includes duckdb.h instead of duplicating everything
- I've reworking the codegen script a little bit. I can be made a little more pretty but it should suffice for now
- the capi_demo extension is still there present, i've left it in so we can test the end-to-end process. We should later move towards migrating an actual extension to the c api then remove this one
- the entrypoint function is now significantly cleaner with a handshake-like procedure. Firstly the DuckDB calls into the extension with a struct of function pointers with some basic functions to get a database pointer and a function pointer struct. During this process the extension can set errors which will then be thrown by DuckDB. As can be seen in `extension/demo_capi/capi_demo.cpp` I've added some macro sugar on top to ensure this is as clean as possible in C land. For extensions written in other languages, the extension templates should probably contain some similar code.

I think we can consider merging this as is, then follow up with:

- A PR to clean up the function comments (e.g. removing unnecessary newlines, improve structure)
- A PR to remove the currently pinned order of the function groups (which was maintained for review purposes, see TODO in `scripts/generate_c_api.py`)
- A PR to add testing infrastructure that tests ALL extension C api functions
- A PR that adds a mechanism to manage which functions end up in the function struct: currently when adding functions to the C API they are not added the the function struct. We should have some procedure thought out with some scripts to make this process smooth. 
- A PR that adds a dev-version of the function pointer struct. To elaborate: we probably don't want to bump the version of the function struct for every PR that introduces a function we want to add. Therefore we might only collect the functions that have been tested and proven to work, before they make in into the function struct since we can not change it. But since we do want to be able to test it, we should probably have some sort of way for an extension to indicate: just give me a unstable struct of all function pointers, which would then be tied to a specific DuckDB version. This would be a way for extensions to use functions that are in the C API but are not yet in the stable struct.

Mytherin commented on CMakeLists.txt:
```diff
@@ -876,11 +880,18 @@ function(build_loadable_extension_directory NAME OUTPUT_DIRECTORY EXTENSION_VERS
       )
   endif()
+
+  if (${ABI_TYPE} STREQUAL "CPP")
```
In a follow-up PR can we rename this to `CPP_DEPRECATED` to indicate we plan to unify this under `C_STRUCT` in the future?

Mytherin commented on extension/demo_capi/capi_demo.cpp:
```diff
@@ -0,0 +1,31 @@
+}
+
+extern "C" {
+DUCKDB_EXTENSION_REGISTER_ENTRYPOINT(demo_capi, Entrypoint, CAPI_VERSION)
```
Maybe we can inline the `extern "C"` and `CAPI_VERSION` here?

For `extern "C"` generally you check if you are running in cpp
```cpp
#ifdef cpp
extern "C"
#endif

```

Mytherin commented on src/include/duckdb/main/capi/header_generation/header_base.hpp:
```diff
@@ -0,0 +1,616 @@
+//! Note that using older API versions (i.e. not using DUCKDB_API_LATEST) is deprecated.
+//! These will not be supported long-term, and will be removed in future versions.
+#ifndef DUCKDB_API_0_3_1
+#define DUCKDB_API_0_3_1 1
```
Let's remove all of this, it's been long enough

Mytherin commented on src/include/duckdb/main/capi/header_generation/apis/extension_api_v0.json:
```diff
@@ -0,0 +1,299 @@
+        {
+            "version": "v0.0.1",
+            "entries": [
+                {"name": "duckdb_open"},
```
nit: can this be an array?

Mytherin commented on src/include/duckdb/main/capi/header_generation/apis/extension_api_v0.json:
```diff
@@ -0,0 +1,299 @@
+            ]
+        },
+        {
+            "version": "v0.0.3",
```
"version": "dev"

This is the unreleased version that contains all added structs

Before a release we modify this to `"version": "x"` and empty the dev version

Mytherin commented on src/include/duckdb/main/capi/header_generation/apis/extension_api_v0.json:
```diff
@@ -0,0 +1,299 @@
+                {"name": "duckdb_append_blob"},
+                {"name": "duckdb_append_null"},
+                {"name": "duckdb_append_data_chunk"},
+                {"name": "duckdb_query_arrow"},
```
These are deprecated, can we remove them?

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -1,11 +1,19 @@
-#include "duckdb/common/serializer/binary_deserializer.hpp"
 #include "mbedtls_wrapper.hpp"
 
+#include <duckdb/function/table/system_functions.hpp>
```
nit: can we remove this or at least fix the quotes?

Mytherin reviewed (commented):
Thanks for the fixes! Some more comments:

samansmink committed:
add some macro magic to make the entrypoint nicer

samansmink committed:
remove old defines from duckdb.h for versioning

samansmink committed:
added exclusion list mechanism, updated readme

samansmink committed:
improve exclusion list validation, remove deprecated functions

samansmink committed:
Merge branch 'main' into c-api-extensions

samansmink committed:
fix merge conflicts of capi changes into generated c api

samansmink committed:
small refactor splitting up the api struct definition files

samansmink committed:
remove nonsensical excluded_from_struct

samansmink committed:
added new split json api definition files

samansmink committed:
clean up generate_c_api script

samansmink committed:
better c ext api versioning working now

samansmink committed:
remove ifdef from typedefs

samansmink committed:
add missing file

samansmink committed:
Merge branch 'main' into c-api-extensions

samansmink committed:
remove accidentally added header

samansmink committed:
add extension entry for demo_capi extension

samansmink committed:
remove unnecessary _init_c_api from macos whitelist for regular extensions

samansmink committed:
update README

samansmink commented:
I've processed the comments here. To summarize the changes:
- removed the `DUCKDB_API_0_3_1` etc. version defines from `duckdb.h` Note that ive also removed the `DUCKDB_NO_EXTENSION_FUNCTIONS` flag.
- refactored the api struct definition file into multiple files to make it more explicit that you can only add new ones, not edit existing ones.
- reworked the way we enforce the versioning of the api struct: this now works using defines that prevent compilation if a function is used from a version higher than the one selected. 
- the API struct is now created on stack and copied into the extension global variable
- there is now a "dev" version which can hold newly added functions, these functions are not promised to be stable and cannot be used by extensions that want binary compatibility with multiple versions of duckdb.
- there is an exclude list for functions that are not in the api struct to make that explicit. This is now enforced by the api generation script
- the api generation script is refactored slightly

samansmink committed:
tidy fix

Mytherin commented on src/include/duckdb_extension.h:
```diff
@@ -0,0 +1,805 @@
+// Function pointer struct
+//===--------------------------------------------------------------------===//
+typedef struct {
+#if DUCKDB_EXTENSION_API_VERSION_MINOR >= 0 && DUCKDB_EXTENSION_API_VERSION_PATCH >= 1 // v0.0.1
```
Major + minor is sufficient for the version info, can we remove the patch?

samansmink committed:
tidy fix

Mytherin commented:
Thanks!

jangorecki commented:
I wonder if we could get a follow up of this PR that would cover
- extension written in C, rather than C++ as now in extension/demo_capi
- readme that describes structure of the new interface

samansmink commented:
Hey @jangorecki yea there will be follow ups here! Both documentation and testing are lacking still.

As things materialize here, I'm also planning to write a blogpost to describe the whole thing in more detail.

a10y commented:
Hey! Just curious, are there any plans to add the filters to the C API? I have a file format (implemented in Rust) that I'd like to write an extension for reading. The format has statistics/zone-maps so it could support predicate pushdown, but AFAICT that's currently not exposed as part of the table function C API.

Mytherin commented:
Yes, that is definitely on the roadmap

samansmink commented:
@Florents-Tselai did you follow the example extension? https://github.com/duckdb/duckdb/tree/main/extension/demo_capi

ddevienne commented:
Hi. The `_extension.h` header is missing from the amalgamation source release. Normal?

samansmink commented:
@ddevienne I guess it should be added there indeed. For now you can just grab the header from https://github.com/duckdb/duckdb/blob/v1.1.1/src/include/duckdb_extension.h

yiyuanliu commented:
How can I determine if a value is null in the C API? It seems there isn't a method for this currently. If I call duckdb_get_xxx methods directly on a null value, it crashes.
